### DISCUSSION TITLE
live grep with exclusions

### DISCUSSION THREAD

**Discussion by kapral18** [ibhagwan/fzf-lua#1194] [Q&A] [answered]
I am using a lot of live_grep with a lot of exclusions like this

![image](https://github.com/ibhagwan/fzf-lua/assets/1625373/3a5c5de8-b0b1-4489-993c-c43969f208dc)

I want to be able to have sets of those exclusions and switch between them while using live_grep

Ideally I'd have a keymap that pops open the picker with prefilled exclusions and puts the cursor right between `\b<cursor>\b`, so I don't have to go back from end of line to type the live_query

I tried passing search opt to live_grep, but it puts the cursor at the end. Can anybody hint at a right direction here?

Thanks

**Comment by ibhagwan** [comment]
There are many ways to go about this problem, one way I can think of is crafting your own `rg_glob_fn` which swaps the args args and query (globs first, query being last) so now you’ll be able to use `<globs> … -- <query>`.
https://github.com/ibhagwan/fzf-lua/wiki#how-can-i-send-custom-flags-to-ripgrep-with-live_grep

> I tried passing search opt to live_grep, but it puts the cursor at the end.

You can send `<C-a>`(using `feedkeys`) in your mapping to send the cursor to the start of line.

Another option is to open `grep` instead (not `live_grep`) with an empty search and the glob prefilled and then performing fuzzy search inside the globs only (as if you already pressed `ctrl-g`):
```lua
:lua require'fzf-lua'.grep({ no_esc=true, rg_glob=true, search=' -- *.lua' })
```

**Reply by kapral18** [reply]
thanks

**Comment by ibhagwan** [comment, marked as answer]
>Ideally I'd have a keymap that pops open the picker with prefilled exclusions and puts the cursor right between \b<cursor>\b

Found your ideal solution using fzf's `start` event:
```lua
:lua require'fzf-lua'.live_grep({
  no_esc=true,
  rg_glob=true, search=[[\b\b -- *.lua]],
  keymap = { fzf = { start = "beginning-of-line+forward-char+forward-char" }}
})
```

We can also use the "dotted" format for a slightly shorter command:
```lua
:lua require'fzf-lua'.live_grep({
  no_esc=true,
  rg_glob=true, search=[[\b\b -- *.lua]],
  ["keymap.fzf.start"] = "beginning-of-line+forward-char+forward-char"
})
```

For more options to send to fzf, `man fzf` and search for `AVAILABLE ACTIONS`:
https://github.com/junegunn/fzf/blob/4e9e842aa4df67051ce88963b0745679dbaeadb5/man/man1/fzf.1#L1296

**Reply by kapral18** [reply]
this is perfection. thank you

**Reply by kapral18** [reply]
@ibhagwan  btw how can I achieve the same for grep_cword and grep_visual? last I checked cword doesn't respond to -- pattern

**Reply by ibhagwan** [reply]
You'd need to craft your own version of `cword`:
> we don't need `keymap.fzf.start` since query is starting empty
```lua
require 'fzf-lua'.grep({
  no_esc = true,
  rg_glob = true,
  search = [[\b]] .. vim.fn.expand("<cword>") .. [[\b -- *.lua]],
})
```

**Reply by ibhagwan** [reply]
For `grep_visual` you can use:
> Note the emission of `no_esc=true` as here **you want to escape special chars**
```lua
require "fzf-lua".grep({
  rg_glob = true,
  search = require("fzf-lua").utils.get_visual_selection() .. [[ -- *.lua]],
})
```

**Reply by kapral18** [reply]
@ibhagwan unrelated followup:

for example if I have this now

```lua
      {
        "<leader>sgg",
        getFzfFn("live_grep", {
          winopts = winopts.large.vertical,
          rg_glob = true,
          no_esc = true,
          search = [[\b\b -- **/*.ts **/*.tsx !*.d.ts !*.test.ts !*.test.tsx !**/__mocks__/* !**/__jest__/* !**/fixtures/* !**/test/* !**/mock/*]],
          keymap = { fzf = { start = "beginning-of-line+forward-char+forward-char" } },
        }),
        desc = "Live Grep (-tests)",
      },
      {
        "<leader>sgt",
        getFzfFn("live_grep", {
          winopts = winopts.large.vertical,
          rg_glob = true,
          no_esc = true,
          search = [[\b\b -- **/*.ts **/*.tsx !*.d.ts]],
          keymap = { fzf = { start = "beginning-of-line+forward-char+forward-char" } },
        }),
        desc = "Live Grep (+tests)",
      },
```

How can I go about activating keymap within a specific keymap to toggle between different search patterns similarly to how ctrl-g is activated within grep? and even before that how can I make patterns specific to buffer, like ts patterns for ts buffer and lua patterns for lua buffer?

**Reply by ibhagwan** [reply]
> how can I make patterns specific to buffer, like ts patterns for ts buffer and lua patterns for lua buffer

You'll need to to create a keymap with a lua callback that sends different fzf-lua commands based on `vim.bo.filetype` (try `:= print(vim.bo.filetype`).

> How can I go about activating keymap within a specific keymap to toggle between different search patterns similarly to how ctrl-g is activated within grep

Add a ctrl-g action and map it to whatever you want:
```lua
  actions = { ["ctrl-g"] = { function() require("fzf-lua").grep({ ... }) end }
```

**Reply by kapral18** [reply]
thanks for the cues

**Reply by kapral18** [reply]
@ibhagwan  I tried this for a cycle through solution, but seems like the nested action only triggers once

```lua
      {
        "<leader>sg",
        function()
          local without_tests =
            [[\b\b -- **/*.ts **/*.tsx !*.d.ts !*.test.ts !*.test.tsx !**/__mocks__/* !**/__jest__/* !**/fixtures/* !**/test/* !**/mock/*]]
          local with_tests = [[\b\b -- **/*.ts **/*.tsx !*.d.ts]]

          local default_search = ""

          local search_list = {
            default_search,
            without_tests,
            with_tests,
          }

          local current_index = 1
          local function get_next_search()
            -- Update the index for the next call
            current_index = (current_index % #search_list) + 1
            return search_list[current_index]
          end

          local live_grep_fn = require("fzf-lua")["live_grep"]
          live_grep_fn({
            winopts = winopts.large.vertical,
            rg_glob = true,
            no_esc = true,
            actions = {
              ["ctrl-g"] = {
                function()
                  live_grep_fn({
                    winopts = winopts.large.vertical,
                    rg_glob = true,
                    no_esc = true,
                    search = get_next_search(),
                    keymap = { fzf = { start = "beginning-of-line+forward-char+forward-char" } },
                  })
                end,
              },
            },
          })
        end,
        desc = "Live Grep",
      },

```

but native ctrl-g seems to take over, I also tried with ctrl-u still doesn't work, it only cycles once. I printed body of action it is only invoked once for whatever reason.

**Reply by ibhagwan** [reply]
> but native ctrl-g seems to take over, I also tried with ctrl-u still doesn't work, it only cycles once. I printed body of action it is only invoked once for whatever reason.

You're only setting `ctrl-g` on the original `live_grep_fn`, when you call the function again (within the `ctrl-g` callback) you aren't setting `actions` so why would it have the action set?

Either make your options a local scope variable or duplicate the `ctrl-g` in the action call as well.

**Reply by kapral18** [reply]
I see I missed the fact that new invocations replace the previous pane, makes sense now, thanks.

**Reply by kapral18** [reply]
@ibhagwan I was trying your suggestion for selection mode
but seems there are 2 problem first is that it says the action is unsupported and the second is that it requires `--multiline` flag (not `--multi` to be passed to fzf otherwise it ignores `\n`), I tried `--multiline` with `fzf_opts` but it crashes with unrecognized flag.

my curr setup for reference:
```lua
local function live_grep_with_patterns(initial_search, opts)
  local search_patterns = {
    initial_search,
    [[\b]]
      .. initial_search
      .. [[\b -- **/*.ts **/*.tsx !*.d.ts !*.test.ts !*.test.tsx !**/__mocks__/* !**/__jest__/* !**/fixtures/* !**/test/* !**/mock/*]],
    [[\b]] .. initial_search .. [[\b -- **/*.ts **/*.tsx !*.d.ts]],
  }
  local current_index = 1

  local function cycle_current_index()
    current_index = (current_index % #search_patterns) + 1
  end

  local function nested_live_grep()
    require("fzf-lua").live_grep(vim.tbl_deep_extend("force", {
      winopts = winopts.large.vertical,
      rg_glob = true,
      no_esc = true,
      actions = {
        ["ctrl-g"] = { nested_live_grep }, -- No need to pass rg_opts here
      },
      search = search_patterns[current_index],
      ["keymap.fzf.start"] = "beginning-of-line+forward-char+forward-char",
    }, opts))

    cycle_current_index()
  end

  nested_live_grep()
end

--...

      {
        "<leader>sv",
        function()
          live_grep_with_patterns(vim.trim(require("fzf-lua").utils.get_visual_selection()), {
            rg_opts = rg_opts,
            no_esc = false,
          })
        end,
        mode = "v",
        desc = "Live Grep Selection",
      },
      {
        "<leader>sV",
        function()
          live_grep_with_patterns(vim.trim(require("fzf-lua").utils.get_visual_selection()), {
            rg_opts = rg_opts_unrestricted,
            no_esc = false,
          })
        end,
        mode = "v",
        desc = "Live Grep Selection (+ignored)",
      },

```

![image](https://github.com/ibhagwan/fzf-lua/assets/1625373/c3e980fc-e59d-4b6a-a362-d004cbd6f09b)

![image](https://github.com/ibhagwan/fzf-lua/assets/1625373/7ec9ccb4-322c-4122-96b0-c108a5671784)

**Reply by ibhagwan** [reply]
> I was trying your suggestion for selection mode
but seems there are 2 problem first is that it says the action is unsupported and the second is that it requires --multiline flag (not --multi to be passed to fzf otherwise it ignores \n), I tried --multiline with fzf_opts but it crashes with unrecognized flag.

You’re sending a new line somewhere as your regex to `rg` which is complaining, multiline is an rg flag not an fzf flag so obviously this won’t work in `fzf_opts`, you can either add it to `rg_opts` as suggested (bad back) or find out why you’re sending a neline to rg in your code.

**Reply by kapral18** [reply]
ok cool the --multiline on rg_opts fixed that. But the subsequent ctrl-g fails with unsupported action even tho it works for all other cases

**Reply by ibhagwan** [reply]
You're missing something somwhere, add `debug=true` and explore the binds, if you have no `ctrl-g` or it's malformed you'll get the error you're describing.

**Reply by kapral18** [reply]
@ibhagwan I debugged a bit and here is what I found:

https://github.com/ibhagwan/fzf-lua/blob/53698d7ff093592228d14a509933ce403b3b8f9f/lua/fzf-lua/core.lua#L211
returns a table with a weird structure when `--multiline` is used, `{last row of selection, action, first row of selection}`

for example in my case 

when I visually select 

![image](https://github.com/ibhagwan/fzf-lua/assets/1625373/f898975b-f68a-4b3e-911a-ec7a43feac98)

and then `ctrl-g` on it in the pane

![image](https://github.com/ibhagwan/fzf-lua/assets/1625373/bb898b5e-b129-423a-806c-4df9a0a8971d)

in fzf-lua core I get this

![image](https://github.com/ibhagwan/fzf-lua/assets/1625373/c33e66eb-e89f-4751-a5b5-cb5d83e17928)

and then `act` is called with that weird table split into 2 arguments, the last row selection goes into `actions` and the actual action (ctrl-g in my case) + first row selection grouped go into `selected` https://github.com/ibhagwan/fzf-lua/blob/53698d7ff093592228d14a509933ce403b3b8f9f/lua/fzf-lua/core.lua#L138

and then that is passed to `normalized_selected` from inside `act` which should return a `keybind`, but instead returns last row of selection as `keybind` hence we get into the remaining else block and see unsupported key error https://github.com/ibhagwan/fzf-lua/blob/53698d7ff093592228d14a509933ce403b3b8f9f/lua/fzf-lua/actions.lua#L77

I guess this means that logic doesn't support multi format yet, can you confirm?

**Reply by ibhagwan** [reply]
> I guess this means that logic doesn't support multi format yet, can you confirm?

I haven’t tested rg multiline so I can’t tell,I don’t see a reason to ever use rg with multiline, I’ll be honest I lost track what you’re trying to achieve but I do recall writing somewhere here that multiline rg is a hack and means you’re doing something wrong somewhere else, bottom line, I don’t think what you’re asking for  is a new functionality, just requires wring the code properly, look at exiting provider code and try to replicate.

**Reply by kapral18** [reply]
@ibhagwan well I am trying to achieve what I have in other IDEs already like jetbrains, where you can search multiline by having a selection of lines and searching them across codebase for a match, for example if you want to see if there is an exact duplication of the lines out there.

**Reply by ibhagwan** [reply]
I don’t understand what is `--multiple`, there is no such flag, there is rg `--multiline` and there is fzf’s `--multi`, there is no connection between these, one tells rg to use multiple lines in the regex and the other dictates fzf to be able to select multiple lines with tab there is no connection between the two.

If there’s a bug open an issue and I’ll fix it otherwise you’re missing something else.

**Reply by kapral18** [reply]
@ibhagwan  oh sorry I misworded I meant `--multiline`, apologies

![image](https://github.com/ibhagwan/fzf-lua/assets/1625373/6d5ad1cc-73e5-40f3-8c92-4039d5644912)

But as I showed above in the screenshot the multiline works and it lists items matched 1 per line and it works as expected for selection but it fails when action is applied against the selection, so this looks indeed like a bug, I just wanted to double check before opening an issue

One more do I understand it correctly that fzf-lua doesn't support native --multi mode?

**Reply by ibhagwan** [reply]
> One more do I understand it correctly that fzf-lua doesn't support native --multi mode?

@kapral18 i absolutely lost you, I have no idea what you’re asking anymore, please post a single focused snippet of code where you believe there’s a bug and I’ll fix it, make the reproduction complete but minimal and tell me why you think something is a bug and how you believe it should behave.

**Reply by kapral18** [reply]
@ibhagwan  Nevermind about the `--multi` then.

I am creating an issue right now. Thank you.

**Comment by dotfrag** [comment]
Hi @ibhagwan, I have a similar question and didn't wanna open a new discussion.

I would like to open `live_grep`, in a specific directory (`cwd`), with a predefined glob pattern, and if possible without showing the glob pattern.

I got this far:

```lua
require("fzf-lua").live_grep({
  cwd = <path>,
  rg_glob = true,
  no_esc = true,
  search = " -- *.lua",
  keymap = { fzf = { start = "beginning-of-line" } },
})
```

Which works great, but I'd like for the pattern to not show in the search, is this possible? I feel that this is answered already but I'm not understanding it.

EDIT: Is this what I want?

```lua
require("fzf-lua").live_grep({
  rg_opts = "-g '*.lua' -e",
})
```

If yes, can I access/use the default `rg_opts` and extend it with `-g ..`?

I think I got it, it works, but is it the right way?

EDIT:

```lua
local rg_opts = require("fzf-lua").defaults.grep.rg_opts
require("fzf-lua").live_grep({
  rg_opts = "-g *.lua " .. rg_opts,
})
```

**Reply by ibhagwan** [reply]
> I think I got it, it works, but is it the right way?

Correct, if you want it not to be part of the query you add it to the rg options.

**Reply by dotfrag** [reply]
Something's off, can you help me troubleshoot?

When specifying `rg_opts` it doesn't return all results. In the same search directory, Searching for `.` (any char), returns 902 results, while without my `rg_opts`, I get 14642.

For example, in my [dotfiles](https://github.com/dotfrag/dotfiles), I get 9 results with `rg_opts` enabled (as posted above), and 12052 without:

![image](https://github.com/ibhagwan/fzf-lua/assets/17456867/b001684a-fd3a-47a1-974b-83a193a3855d)

![image](https://github.com/ibhagwan/fzf-lua/assets/17456867/61a0bbba-c7ba-4672-b703-919e0c101c29)

**Reply by ibhagwan** [reply]
Add `debug = true` to your options and then look at the underlying command and see if some option isn’t right, if everything is fine copy the command to the shell and troubleshoot why it doesn’t return all the results you need.

**Reply by dotfrag** [reply]
Okay, I should have quoted the glob:

`rg_opts = "-g '*.lua' " .. rg_opts,`

Last question, when using `debug=true`, it's difficult to read long lines from the results pane (left), is there any way to scroll? My workaround was to zoom my terminal out to tiny text so I could copy-paste the result (debug line).

**Reply by ibhagwan** [reply]
No way to scroll in fzf, best I know is `--keep-right` but then you’re missing the left side, there might be a solution if I split to multiple lines (based on screen width).

**Reply by dotfrag** [reply]
Okay no problem, thanks for your help!
